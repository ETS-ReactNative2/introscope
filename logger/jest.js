if (
    global.expect &&
    global.expect.addSnapshotSerializer &&
    !process.env.__EFFECTS_LOGGER__NO_JEST_INTERGATION
) {
    const { spySnapshotSerializer } = require('./spySnapshotSerializer');
    const {
        effectsLogSnapshotSerializer,
    } = require('./effectsLogSnapshotSerializer');

    global.expect.addSnapshotSerializer(spySnapshotSerializer);
    global.expect.addSnapshotSerializer(effectsLogSnapshotSerializer);
}
