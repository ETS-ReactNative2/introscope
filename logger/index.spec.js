require('./jest');

const introscope = (scope = {}) => {
    scope.log = (...args) => console.log(...args);
    scope.count = 0;
    scope.newTodo = (id, title) => {
        (0, scope.log)('new todo created', id);
        return {
            id,
            title,
        };
    };
    scope.addTodo = title => (0, scope.newTodo)(++scope.count, title);

    return scope;
};

import { effectsLogger, SPY, KEEP } from '.';
const effectsScope = effectsLogger(introscope);

describe('todos', () => {
    it('addTodo', () => {
        const { effects, addTodo } = effectsScope({
            newTodo: SPY,
            addTodo: KEEP,
        });

        addTodo('start using Introscope :)');

        expect(effects()).toMatchSnapshot();
    });
});

describe('effectsLogger', () => {
    it('adds new props to scope', () => {
        const scope = effectsLogger(() => ({
            a: 0,
        }))({
            b: 2,
        });

        expect(scope).toMatchSnapshot();
    });

    it('does not mock primitives', () => {
        const scope = effectsLogger(() => ({
            undefined: undefined,
            null: null,
            boolean: true,
            number: 1,
            string: 'foo',
            symbol: Symbol('bar'),
        }))({});

        expect(scope).toMatchSnapshot();
    });

    it('auto mocks objects, arrays and functions', () => {
        const scope = effectsLogger(() => ({
            function: function foo() {},
            object: { bar: true },
            array: [1, 2, 3],
        }))({});

        expect(scope).toMatchSnapshot();
    });

    it('has working functionMocker as effects.fn', () => {
        const scope = effectsLogger(() => ({}))({});

        expect(scope.effects.fn.foo()).toMatchSnapshot();
    });
});

import { functionMocker } from '.';

describe('functionMocker', () => {
    it('mocks', () => {
        const m = functionMocker();
        expect(m.foo()).toMatchSnapshot();
        expect(m.bar()).toMatchSnapshot();
    });

    it('logs', () => {
        const m = functionMocker();
        m.baz()(123);
        expect(m()).toMatchSnapshot();
    });

    it('calls and passes args', () => {
        const m = functionMocker();
        m.baz(m.forBaz())(1, 2, 3);
        expect(m()).toMatchSnapshot();
    });

    it('returns value', () => {
        const m = functionMocker();
        expect(m.baz(() => 'returned value')()).toMatchSnapshot();
    });

    it('keeps order', () => {
        const m = functionMocker();
        m.fn1()();
        m.fn2()();
        m.fn3()();
        expect(m()).toMatchSnapshot();
    });

    it('ignores never called mocks', () => {
        const m = functionMocker();
        m.fn1()();
        m.fn2();
        m.fn3()();
        expect(m()).toMatchSnapshot();
    });
});
