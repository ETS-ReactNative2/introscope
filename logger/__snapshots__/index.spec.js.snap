// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effectsLogger adds new props to scope 1`] = `
Object {
  "a": 0,
  "b": 2,
}
`;

exports[`effectsLogger auto mocks objects, arrays and functions 1`] = `
Object {
  "array": Array [],
  "function": [Function],
  "object": Object {},
}
`;

exports[`effectsLogger does not mock primitives 1`] = `
Object {
  "boolean": true,
  "null": null,
  "number": 1,
  "string": "foo",
  "symbol": Symbol(bar),
  "undefined": undefined,
}
`;

exports[`effectsLogger has working functionMocker as effects.fn 1`] = `[Function]`;

exports[`functionMocker calls and passes args 1`] = `
EffectsLog [
  baz(
    1,
    2,
    3,
  ),
  forBaz(
    1,
    2,
    3,
  ),
]
`;

exports[`functionMocker ignores never called mocks 1`] = `
EffectsLog [
  fn1(
  ),
  fn3(
  ),
]
`;

exports[`functionMocker keeps order 1`] = `
EffectsLog [
  fn1(
  ),
  fn2(
  ),
  fn3(
  ),
]
`;

exports[`functionMocker logs 1`] = `
EffectsLog [
  baz(
    123,
  ),
]
`;

exports[`functionMocker mocks 1`] = `[Function]`;

exports[`functionMocker mocks 2`] = `[Function]`;

exports[`functionMocker returns value 1`] = `"returned value"`;

exports[`todos addTodo 1`] = `
EffectsLog [
  module.count =
    1,
  newTodo(
    1,
    "start using Introscope :)",
  ),
  log(
    "new todo created",
    1,
  ),
]
`;
